src/basedtyping/__init__.py

```diff
@@ -591,6 +587,116 @@ def deco(fn: Callable[[], None]) -> Callable[[], None]: ...
         @deco
         def foo(): ...
     """
-    if not isinstance(fn, FunctionType):  # type: ignore[redundant-expr]
+    if not isinstance(fn, FunctionType):
         raise TypeError(f"{fn} is not a FunctionType")
-    return fn  # type: ignore[unreachable]
+    # https://github.com/KotlinIsland/basedmypy/issues/745
+    return cast("FunctionType[P, T]", fn)
+
+
+class ForwardRef(typing.ForwardRef, _root=True):  # type: ignore[call-arg,misc]
+    """
+    Like `typing.ForwardRef`, but lets older Python versions use newer typing features.
+    Specifically, when evaluated, this transforms `X | Y` into `typing.Union[X, Y]`
+    and `list[X]` into `typing.List[X]` etc. (for all the types made generic in PEP 585)
+    if the original syntax is not supported in the current Python version.
+    """
+
+    # older typing.ForwardRef doesn't have this
+    __slots__ = ["__forward_module__", "__forward_is_class__"]
+
+    def __init__(self, arg: str, *, is_argument=True, module: object = None, is_class=False):
+        if not isinstance(arg, str):  # type: ignore[redundant-expr]
+            raise TypeError(f"Forward reference must be a string -- got {arg!r}")
+
+        # If we do `def f(*args: *Ts)`, then we'll have `arg = '*Ts'`.
+        # Unfortunately, this isn't a valid expression on its own, so we
+        # do the unpacking manually.
+        arg_to_compile = (
+            f"({arg},)[0]"  # E.g. (*Ts,)[0] or (*tuple[int, int],)[0]
+            if arg.startswith("*")
+            else arg
+        )
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=SyntaxWarning)
+                code = compile(arg_to_compile, "<string>", "eval")
+        except SyntaxError:
+            try:
+                ast.parse(arg_to_compile, mode="func_type")
+            except SyntaxError:
+                raise SyntaxError(f"invalid syntax in ForwardRef: {arg_to_compile}?") from None
+            else:
+                code = compile("'un-representable callable type'", "<string>", "eval")
+
+        self.__forward_arg__ = arg
+        self.__forward_code__ = code
+        self.__forward_evaluated__ = False
+        self.__forward_value__ = None
+        self.__forward_is_argument__ = is_argument
+        self.__forward_is_class__ = is_class
+        self.__forward_module__ = module
+
+    if sys.version_info >= (3, 13):
+
+        @override
+        def _evaluate(
+            self,
+            globalns: dict[str, object] | None,
+            localns: dict[str, object] | None,
+            type_params: tuple[TypeVar | typing.ParamSpec | typing.TypeVarTuple, ...] = (),
```

Comment:
This wouldn't allow the `typing_extensions` variants for older versions (assuming that you're gonna support it)

Reply:
Either way, this could be made more DRY with a `TypeAlias`

Reply:
i think this will work, these names are imported from `typing_extensions`, so they will always work?